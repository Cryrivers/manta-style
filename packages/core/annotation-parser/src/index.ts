import { AnnotationAst, parseAnnotationFromString } from './parser';

export { AnnotationAst, parseAnnotationFromString };
export const JsdocKey = 'mantastyle';

const MANTA_STYLE_ANNOTATION_REGEX = new RegExp(
  `@${JsdocKey}\\s+({{[\\s\\S]+(?=}})}})`,
);

export function extractMantaStyleJSDocContent(
  multilineComment: string,
): string {
  const match = multilineComment.match(MANTA_STYLE_ANNOTATION_REGEX);
  if (match) {
    return cleanMantaStyleJSDocContent(match[1]);
  }
  return '';
}

export function cleanMantaStyleJSDocContent(multilineContent: string): string {
  multilineContent = multilineContent.trimLeft();
  if (multilineContent.startsWith('{{')) {
    let bracketEnd = 2;
    let quoteMode = false;
    let singleQuote = false;
    loop: while (true) {
      switch (multilineContent[bracketEnd]) {
        case '\\':
          bracketEnd++;
          break;
        case '}': {
          if (!quoteMode && multilineContent[bracketEnd + 1] === '}') {
            break loop;
          }
        }
        case '"':
        case "'":
          if (quoteMode) {
            if (
              (singleQuote && multilineContent[bracketEnd] === "'") ||
              (!singleQuote && multilineContent[bracketEnd] === '"')
            ) {
              quoteMode = false;
            }
          } else {
            quoteMode = true;
            singleQuote = multilineContent[bracketEnd] === "'";
          }
      }
      bracketEnd++;
    }
    return multilineContent.substring(0, bracketEnd + 2);
  }
  return '';
}
