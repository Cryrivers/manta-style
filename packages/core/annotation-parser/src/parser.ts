import { parse } from 'handlebars';

type AnnotationLiteral = {
  type: 'literal';
  value: string | number | boolean | undefined | null;
};

type AnnotationExpression = {
  type: 'expression';
  name: string;
  params: Array<AnnotationNode>;
  hash: { [name: string]: AnnotationNode };
};

type AnnotationNode = AnnotationLiteral | AnnotationExpression | undefined;

export type AnnotationAst = AnnotationNode;

export function parseAnnotationFromString(
  handlebarString: string,
): AnnotationAst {
  const mustacheStatement = parse(handlebarString).body[0];
  return simplifyAst(mustacheStatement);
}

function simplifyAst(statement: hbs.AST.Expression | null): AnnotationNode {
  if (!statement) {
    return undefined;
  }
  switch (statement.type) {
    case 'MustacheStatement':
    case 'SubExpression': {
      // @ts-ignore
      const fnName: string = statement.path.original;
      const params: Array<AnnotationNode> = [];
      const hash: { [name: string]: AnnotationNode } = {};
      // @ts-ignore
      for (const param of statement.params) {
        params.push(simplifyAst(param));
      }
      // @ts-ignore
      if (statement.hash) {
        // @ts-ignore
        for (const hashPair of statement.hash.pairs) {
          // @ts-ignore
          hash[hashPair.key] = simplifyAst(hashPair.value);
        }
      }
      return {
        type: 'expression',
        name: fnName,
        params,
        hash,
      };
    }
    case 'PathExpression':
      return {
        type: 'expression',
        // @ts-ignore
        name: statement.original,
        params: [],
        hash: {},
      };
    case 'StringLiteral':
    case 'BooleanLiteral':
    case 'NumberLiteral':
      return {
        type: 'literal',
        // @ts-ignore
        value: statement.value,
      };
    case 'UndefinedLiteral':
      return {
        type: 'literal',
        value: undefined,
      };
    case 'NullLiteral':
      return {
        type: 'literal',
        value: null,
      };
  }
  return undefined;
}
