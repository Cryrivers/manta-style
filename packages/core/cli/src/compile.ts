#!/usr/bin/env node
import * as path from 'path';
import * as builder from '@manta-style/typescript-builder';
import * as program from 'commander';
import findRoot = require('find-root');
import chalk from 'chalk';

program
  .version('0.0.11')
  .option(
    '-c --configFile <file>',
    'the TypeScript config file to generate entry points',
  )
  .option('-v --verbose', 'show debug information')
  .parse(process.argv);

const { configFile, verbose = false } = program;

if (!configFile) {
  console.log(
    'Please specifiy a entry point config file by using --configFile.',
  );
  process.exit(1);
}

const tmpDir = findRoot(process.cwd()) + '/.mantastyle-tmp';
builder.build(path.resolve(configFile), tmpDir, verbose);
console.log(
  chalk.green(
    `Built manta-style runnable code into ${chalk.underline.white(tmpDir)}`,
  ),
);
