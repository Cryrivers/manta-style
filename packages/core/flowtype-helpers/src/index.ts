export type $Keys<T> = keyof T;

export type $Values<T> = T[keyof T];

export type $PropertyType<T, K extends keyof T> = { [P in K]: T[P] };

export type $Shape<T> = { [P in keyof T]?: T[P] };

export type $ReadOnly<T> = T;

export type $Exact<T> = T;

export type $Diff<A, B> = A extends B ? never : A;

export type $Rest<A, B> = A extends B ? never : (A | undefined);

export type $NonMaybeType<T> = T extends null | undefined ? never : T;

export type Array<T> = T[];

export type $ReadOnlyArray<T> = T[];
