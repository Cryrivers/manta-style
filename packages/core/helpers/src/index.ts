import QueryType from './types/QueryType';
import MantaStyle from '@manta-style/runtime';

export const Query = MantaStyle.TypeAliasDeclaration(
  'Query',
  (typeFactory) => {
    const T = typeFactory.TypeParameter('T');
    return new QueryType(T);
  },
  [],
);
