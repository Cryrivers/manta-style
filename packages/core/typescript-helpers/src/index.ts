/**
 * Make all properties in T optional
 * @preserveUnion
 */
export type Partial<T> = { [P in keyof T]?: T[P] };

/**
 * Make all properties in T required
 * @preserveUnion
 */
export type Required<T> = { [P in keyof T]-?: T[P] };

/**
 * Make all properties in T readonly
 * @preserveUnion
 */
export type Readonly<T> = { readonly [P in keyof T]: T[P] };

/**
 * From T pick a set of properties K
 * @preserveUnion
 */
export type Pick<T, K extends keyof T> = { [P in K]: T[P] };

/**
 * Construct a type with a set of properties K of type T
 * @preserveUnion
 */
export type Record<K extends keyof any, T> = { [P in K]: T };

/**
 * Exclude from T those export types that are assignable to U
 * @preserveUnion
 */
export type Exclude<T, U> = T extends U ? never : T;

/**
 * Extract from T those export types that are assignable to U
 * @preserveUnion
 */
export type Extract<T, U> = T extends U ? T : never;

/**
 * Exclude null and undefined from T
 * @preserveUnion
 */
export type NonNullable<T> = T extends null | undefined ? never : T;
