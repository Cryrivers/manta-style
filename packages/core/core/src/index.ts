import { MantaStyleAnnotation } from './annotation';
import { AnnotationAst } from '@manta-style/annotation-parser';
import { PluginSystem } from './plugin';
import { Annotation } from './utils/annotation';

export type MantaStyleContext = {
  query: { [key: string]: unknown };
  plugins: PluginSystem;
};
export * from './plugin';
export { MantaStyleAnnotation, AnnotationAst, Annotation };

export abstract class Type {
  abstract deriveLiteral(
    parentAnnotations: MantaStyleAnnotation,
    context: MantaStyleContext,
  ): Promise<Type>;
  public mock(): any {
    throw new Error('Literal types should be derived before mock.');
  }
}
