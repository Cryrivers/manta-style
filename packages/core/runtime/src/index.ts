import NumberKeyword from './types/NumberKeyword';
import BooleanKeyword from './types/BooleanKeyword';
import UndefinedKeyword from './types/UndefinedKeyword';
import NullKeyword from './types/NullKeyword';
import TypeLiteral from './types/TypeLiteral';
import ArrayType from './types/ArrayType';
import Literal from './types/Literal';
import UnionType from './types/UnionType';
import AnyKeyword from './types/AnyKeyword';
import NeverKeyword from './types/NeverKeyword';
import StringKeyword from './types/StringKeyword';
import TypeAliasDeclaration from './nodes/TypeAliasDeclaration';
import LazyTypeAliasDeclaration from './nodes/LazyTypeAliasDeclaration';
import ConditionalType from './types/ConditionalType';
import KeyOfKeyword from './types/KeyOfKeyword';
import TupleType from './types/TupleType';
import RestType from './types/RestType';
import OptionalType from './types/OptionalType';
import IndexedAccessType from './types/IndexedAccessType';
import MappedType from './types/MappedType';
import { Literals, Property } from './utils/baseType';
import IntersectionType from './types/IntersectionType';
import ParenthesizedType from './types/ParenthesizedType';
import ObjectKeyword from './types/ObjectKeyword';
import ArrayLiteral from './types/ArrayLiteral';
import { Annotation, Type } from '@manta-style/core';

export type TypeAliasDeclarationFactory = () => TypeAliasDeclaration;
export type TypeLiteral = TypeLiteral;
export type Property = Property;
class MantaStyle {
  public static TypeAliasDeclaration(
    typeName: string,
    typeCallback: (currentType: TypeAliasDeclaration) => Type,
    annotations: Annotation[],
  ) {
    const newType = new LazyTypeAliasDeclaration(typeName, annotations);
    newType.setInitialize(typeCallback);
    return newType;
  }
  public static TypeLiteral(typeCallback: (currentType: TypeLiteral) => void) {
    const newType = new TypeLiteral();
    typeCallback(newType);
    return newType;
  }
  public static MappedType(typeCallback: (currentType: MappedType) => Type) {
    const newType = new MappedType();
    newType.setType(typeCallback(newType));
    return newType;
  }
  public static IndexedAccessType(objectType: Type, indexType: Type) {
    return new IndexedAccessType(objectType, indexType);
  }
  public static UnionType(types: Type[]) {
    return new UnionType(types);
  }
  public static ParenthesizedType(type: Type) {
    return new ParenthesizedType(type);
  }
  public static IntersectionType(types: Type[]) {
    return new IntersectionType(types);
  }
  public static Literal(literal: Literals) {
    return new Literal(literal);
  }
  public static ArrayType(elementType: Type) {
    return new ArrayType(elementType);
  }
  public static ArrayLiteral(elements: Type[]) {
    return new ArrayLiteral(elements);
  }
  public static TupleType(elementTypes: Type[]) {
    return new TupleType(elementTypes);
  }
  public static RestType(elementType: Type) {
    return new RestType(elementType);
  }
  public static OptionalType(type: Type) {
    return new OptionalType(type);
  }
  public static ConditionalType(
    checkType: Type,
    extendsType: Type,
    trueType: Type,
    falseType: Type,
  ) {
    return new ConditionalType(checkType, extendsType, trueType, falseType);
  }
  public static NumberKeyword = new NumberKeyword();
  public static BooleanKeyword = new BooleanKeyword();
  public static StringKeyword = new StringKeyword();
  public static NeverKeyword = new NeverKeyword();
  public static NullKeyword = new NullKeyword();
  public static UndefinedKeyword = new UndefinedKeyword();
  public static AnyKeyword = new AnyKeyword();
  public static ObjectKeyword = new ObjectKeyword();
  public static KeyOfKeyword(type: Type) {
    return new KeyOfKeyword(type);
  }
}

export default MantaStyle;
export const LiteralType = Literal;
export { resolveReferencedType } from './utils/referenceTypes';
