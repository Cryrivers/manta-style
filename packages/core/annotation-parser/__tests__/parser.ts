import { parseAnnotationFromString } from '../src/index';

describe('Annotation', () => {
  test('simple', () => {
    expect(prettyParse('{{ plugin }}')).toMatchSnapshot();
    expect(prettyParse('{{ plugin "abc" 123 true }}')).toMatchSnapshot();
    expect(
      prettyParse('{{ plugin "abc" 123 true d="123" e=12 f=false }}'),
    ).toMatchSnapshot();
  });
  test('nested', () => {
    expect(prettyParse('{{ plugin "abc" nested }}')).toMatchSnapshot();
    expect(
      prettyParse('{{ plugin "abc" (nested "def") 123 }}'),
    ).toMatchSnapshot();
  });
});

function prettyParse(annotation: string) {
  return {
    FROM: annotation,
    TO: parseAnnotationFromString(annotation),
  };
}
