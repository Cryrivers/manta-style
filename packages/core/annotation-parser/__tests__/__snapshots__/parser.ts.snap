// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Annotation nested 1`] = `
Object {
  "FROM": "{{ plugin \\"abc\\" nested }}",
  "TO": Object {
    "hash": Object {},
    "name": "plugin",
    "params": Array [
      Object {
        "type": "literal",
        "value": "abc",
      },
      Object {
        "hash": Object {},
        "name": "nested",
        "params": Array [],
        "type": "expression",
      },
    ],
    "type": "expression",
  },
}
`;

exports[`Annotation nested 2`] = `
Object {
  "FROM": "{{ plugin \\"abc\\" (nested \\"def\\") 123 }}",
  "TO": Object {
    "hash": Object {},
    "name": "plugin",
    "params": Array [
      Object {
        "type": "literal",
        "value": "abc",
      },
      Object {
        "hash": Object {},
        "name": "nested",
        "params": Array [
          Object {
            "type": "literal",
            "value": "def",
          },
        ],
        "type": "expression",
      },
      Object {
        "type": "literal",
        "value": 123,
      },
    ],
    "type": "expression",
  },
}
`;

exports[`Annotation simple 1`] = `
Object {
  "FROM": "{{ plugin }}",
  "TO": Object {
    "hash": Object {},
    "name": "plugin",
    "params": Array [],
    "type": "expression",
  },
}
`;

exports[`Annotation simple 2`] = `
Object {
  "FROM": "{{ plugin \\"abc\\" 123 true }}",
  "TO": Object {
    "hash": Object {},
    "name": "plugin",
    "params": Array [
      Object {
        "type": "literal",
        "value": "abc",
      },
      Object {
        "type": "literal",
        "value": 123,
      },
      Object {
        "type": "literal",
        "value": true,
      },
    ],
    "type": "expression",
  },
}
`;

exports[`Annotation simple 3`] = `
Object {
  "FROM": "{{ plugin \\"abc\\" 123 true d=\\"123\\" e=12 f=false }}",
  "TO": Object {
    "hash": Object {
      "d": Object {
        "type": "literal",
        "value": "123",
      },
      "e": Object {
        "type": "literal",
        "value": 12,
      },
      "f": Object {
        "type": "literal",
        "value": false,
      },
    },
    "name": "plugin",
    "params": Array [
      Object {
        "type": "literal",
        "value": "abc",
      },
      Object {
        "type": "literal",
        "value": 123,
      },
      Object {
        "type": "literal",
        "value": true,
      },
    ],
    "type": "expression",
  },
}
`;
